# Week 5 Homework

>For this homework we will be using the FHVHV 2021-06 data found here. [FHVHV Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz)

In [14]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz -O ./data/fhvhv_2021-06.csv.gz

--2023-03-05 07:56:32--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230305%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230305T075632Z&X-Amz-Expires=300&X-Amz-Signature=820688a9a8d437d4b3048401f04983835bc2c56766f6be81a632680fc42f604c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [following]
--2023-03-05 07:56:32--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e

## Question 1: 

>**Install Spark and PySpark** 
>- Install Spark
>- Run PySpark
>- Create a local spark session
>- Execute spark.version.
>
>What's the output?
- 3.3.2

In [3]:
import pyspark
from pyspark.sql import SparkSession, types

spark = SparkSession.builder \
    .master('local[*]') \
    .appName('test') \
    .getOrCreate()
    
spark.version

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/03/05 11:21:40 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/03/05 11:21:41 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.


'3.3.2'

## Question 2: 

>**HVFHW June 2021**
>
>Read it with Spark using the same schema as we did in the lessons.</br> 
>We will use this dataset for all the remaining questions.</br>
>Repartition it to 12 partitions and save it to parquet.</br>
>What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>

- 24MB

In [17]:
import pandas as pd
df_pd = pd.read_csv('./data/fhvhv_2021-06.csv.gz', \
    nrows = 100)

schema = df_pd.dtypes
schema

dispatching_base_num      object
pickup_datetime           object
dropoff_datetime          object
PULocationID               int64
DOLocationID               int64
SR_Flag                   object
Affiliated_base_number    object
dtype: object

In [5]:
from pyspark.sql import types

schema = types.StructType([
types.StructField('dispatching_base_num', types.StringType(), True), 
types.StructField('pickup_datetime', types.TimestampType(), True), 
types.StructField('dropoff_datetime', types.TimestampType(), True), 
types.StructField('PULocationID', types.LongType(), True), 
types.StructField('DOLocationID', types.LongType(), True), 
types.StructField('SR_Flag', types.StringType(), True),
types.StructField('Affiliated_base_number', types.StringType(), True), 
])

df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv("./data/fhvhv_2021-06.csv.gz")
df.show(5)

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02764|2021-06-01 00:02:41|2021-06-01 00:07:46|         174|          18|      N|                B02764|
|              B02764|2021-06-01 00:16:16|2021-06-01 00:21:14|          32|         254|      N|                B02764|
|              B02764|2021-06-01 00:27:01|2021-06-01 00:42:11|         240|         127|      N|                B02764|
|              B02764|2021-06-01 00:46:08|2021-06-01 00:53:45|         127|         235|      N|                B02764|
|              B02510|2021-06-01 00:45:42|2021-06-01 01:03:33|         144|         146|      N|                  null|
+--------------------+------------------

In [19]:
df = df.repartition(12)
df.write.parquet('./data/fhv_pq')
!ls -lh ./data/fhv_pq

total 284M
-rw-r--r-- 1 sam sam   0 Mar  5 08:03 _SUCCESS
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00000-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00001-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00002-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00003-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00004-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00005-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00006-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00007-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.snappy.parquet
-rw-r--r-- 1 sam sam 24M Mar  5 08:03 part-00008-a58b5092-2344-4cab-9f0a-64a8129e337a-c000.sna

## Question 3: 

>**Count records**  
>
>How many taxi trips were there on June 15?</br></br>
>Consider only trips that started on June 15.</br>

- 452,470

In [20]:
df.registerTempTable('fhvhv')

/home/sam/anaconda3/envs/pyspark/lib/python3.9/site-packages/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [23]:
spark.sql("""
    SELECT
        COUNT(*) AS trips
    FROM fhvhv
    WHERE pickup_datetime >= '2021-06-15'
        AND pickup_datetime < '2021-06-16'
""").show()

+------+
| trips|
+------+
|452470|
+------+



## Question 4: 

>**Longest trip for each day**  
>
>Now calculate the duration for each trip.</br>
>How long was the longest trip in Hours?</br>

- 66.87 Hours

In [7]:
# Trying this with a function, rather than straight SQL
from pyspark.sql import functions as F

df_dur = df.withColumn('TimeDiff', \
    (F.col('dropoff_datetime').cast('long') - \
    F.col('pickup_datetime').cast('long'))/3600)

df_dur.registerTempTable('fhvhv_dur')

/home/sam/anaconda3/envs/pyspark/lib/python3.9/site-packages/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [8]:
spark.sql("""
    SELECT
        dispatching_base_num,
        pickup_datetime,
        dropoff_datetime,
        ROUND(TimeDiff, 2) AS TimeDiff
    FROM fhvhv_dur
    ORDER BY TimeDiff DESC
""").show(10)

+--------------------+-------------------+-------------------+--------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|TimeDiff|
+--------------------+-------------------+-------------------+--------+
|              B02872|2021-06-25 13:55:41|2021-06-28 08:48:25|   66.88|
|              B02765|2021-06-22 12:09:45|2021-06-23 13:42:44|   25.55|
|              B02879|2021-06-27 10:32:29|2021-06-28 06:31:20|   19.98|
|              B02800|2021-06-26 22:37:11|2021-06-27 16:49:01|    18.2|
|              B02682|2021-06-23 20:40:43|2021-06-24 13:08:44|   16.47|
|              B02869|2021-06-23 22:03:31|2021-06-24 12:19:39|   14.27|
|              B02877|2021-06-24 23:11:00|2021-06-25 13:05:35|   13.91|
|              B02765|2021-06-04 20:56:02|2021-06-05 08:36:14|   11.67|
|              B02617|2021-06-27 07:45:19|2021-06-27 19:07:16|   11.37|
|              B02880|2021-06-20 17:05:12|2021-06-21 04:04:16|   10.98|
+--------------------+-------------------+-------------------+--

## Question 5: 

>**User Interface**
>
>Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 4040

### Question 6: 

>**Most frequent pickup location zone**
>
>Load the zone lookup data into a temp view in Spark</br>
>[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>
>
>Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- Crown Heights North

In [34]:
df_zones = spark.read.csv('./data/raw/taxi_zone_lookup.csv', header=True)
df.join(df_zones, df.PULocationID == df_zones.LocationID, how='inner') \
    .registerTempTable('df_join')

/home/sam/anaconda3/envs/pyspark/lib/python3.9/site-packages/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [36]:
spark.sql("""
    SELECT
        COUNT(*) AS trips,
        Zone
    FROM df_join
    GROUP BY Zone
    ORDER BY trips DESC
""").show(10)

+------+--------------------+
| trips|                Zone|
+------+--------------------+
|231279| Crown Heights North|
|221244|        East Village|
|188867|         JFK Airport|
|187929|      Bushwick South|
|186780|       East New York|
|164344|TriBeCa/Civic Center|
|161596|   LaGuardia Airport|
|158937|            Union Sq|
|154698|        West Village|
|152493|             Astoria|
+------+--------------------+
only showing top 10 rows

